### PR TITLE
Special characters provided shouldn't be empty

### PR DESCRIPTION
**Pull request by sharidas**
The checkbox should not allow special characters
to be empty.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

### CODE REVIEW

**Comment by phil-davis** [comment]
If I go to Settings->Admin->Security and check "1 special characters" but do not check "Restrict to these special characters" then when I try to save it says:
```
Error: The special characters cannot be empty.
```
But actually this should work.

When "Restrict to these special characters" is not checked, then the text box can be empty. And actually the system implements a default set of special characters - which is all non-alphanumeric characters.

Also, if I do not even check "1 special characters" then I still get the error message.

**Comment by sharidas** [comment]
## Description
When the user checks `Restrict to these special characters:` and provides empty data, the user creation fails. And hence this change set takes care of the same. I have made changes to the controller as well as the frontend. In the front end care has been taken so that the failure notification would be shown in the UI when the user provides emtpy data to `Restrict to these special characters:` when the checkbox before the label is checked. If it is not checked then notification will not be seen.

## Issue
- Fixes https://github.com/owncloud/core/issues/35743

## Test done
- Check the `Restrict to these special characters:` and provide emtpy data and click the save button here is the screenshot what user would see: 
pp
- Refresh the page and the previous value in the text box will be available.
- If the checkbox for `Restrict to these special characters:` is not checked no failure or notification is seen in the page: 
pp2
- The change notification or failure is dependant on 2 factors :
   - If the checkbox `Restrict to these special characters` is checked and
   - If the input text for `Restrict to these special characters` is empty

File: lib/Controller/SettingsController.php [hunk 1]
```diff
@@ -92,6 +93,9 @@ public function updatePolicy() {
 						$value = $this->$convertFuncName($value);
 					}
 					$this->config->setAppValue('password_policy', $key, $value);
+				} elseif (($this->request->getParam('spv_def_special_chars_checked') === 'on') &&
+					($key === 'spv_def_special_chars_value') && ($this->request->getParam($key) === '')) {
+					throw new PolicyException("The restricetd special characters cannot be empty");
```
Comment (phil-davis) [inline lib/Controller/SettingsController.php]:
```suggestion
					throw new PolicyException("The restricted special characters cannot be empty");
```

Comment (sharidas) [reply lib/Controller/SettingsController.php]: Done. Corrected to `restricted`.

File: js/ajax.js [hunk 2]
```diff
@@ -22,6 +22,11 @@ $(document).ready(function() {
 		event.preventDefault();
 		OC.msg.startSaving('#password_policy .msg');
 
+		if (($("input[name=spv_def_special_chars_checked]").is(':checked') === true) &&
+			($("input[name=spv_def_special_chars_value]").val() === "")) {
+			OC.Notification.showTemporary(t('password_policy', 'Error: The special characters cannot be empty.'));
```
Comment (karakayasemi) [inline js/ajax.js]:
````suggestion
			OC.Notification.showTemporary(t('password_policy', 'Error: The special characters cannot be empty.'));
 			OC.msg.finishedSaving('#password_policy .msg', {status: "failed", data: { message: t('password_policy', 'Failed to save!')} });
 			return;
````
If we can detect an error in client side, I guess no need to post erroneous request to server.  
Also, currently, we are saving `spv_def_special_chars_checked` config value even we say `Failed to save`. Suggested change prevents this situation also .

Comment (sharidas) [reply js/ajax.js]: I was trying to be in the defensive side. In either way the data shouldn't be saved, be it from the client side post or any hackish thing which reaches the server.

Comment (sharidas) [reply js/ajax.js]: Made changes as discussed.

**Comment by phil-davis** [comment]
Good stuff. I tried to break this just now, and I can't break it.
